File: Deber 2/ejercicio_5/mc_solver.py
# Problema de los Misioneros y Canibales - Backtracking (DFS)
# Estado: (M_izq, C_izq, Bote) con Bote: 0=izquierda, 1=derecha. Total: 3 y 3.

START = (3, 3, 0)
GOAL = (0, 0, 1)

MOVES = [  # (misioneros, canibales) que se suben al bote
    (2, 0), (1, 0),
    (0, 2), (0, 1),
    (1, 1),
]


def is_safe(m_left: int, c_left: int) -> bool:
    # Rango valido
    if not (0 <= m_left <= 3 and 0 <= c_left <= 3):
        return False
    # Izquierda
    if m_left > 0 and c_left > m_left:
        return False
    # Derecha
    m_right = 3 - m_left
    c_right = 3 - c_left
    if m_right > 0 and c_right > m_right:
        return False
    return True


def neighbors(state):
    m_left, c_left, boat = state
    for m, c in MOVES:
        # El bote debe llevar max 2 y al menos 1 persona (siempre cierto por nuestra lista)
        if boat == 0:
            # Mover de izquierda a derecha
            nm, nc = m_left - m, c_left - c
        else:
            # Mover de derecha a izquierda
            nm, nc = m_left + m, c_left + c
        if not is_safe(nm, nc):
            continue
        yield (nm, nc, 1 - boat), (m, c)


# --- Backtracking (DFS) ---

def dfs_backtracking(start=START, goal=GOAL):
    visited = {start}
    path = [(start, None)]  # (estado, movimiento)
    return _dfs(start, goal, visited, path)


def _dfs(state, goal, visited, path):
    if state == goal:
        return path
    for ns, move in neighbors(state):
        if ns in visited:
            continue
        visited.add(ns)
        res = _dfs(ns, goal, visited, path + [(ns, move)])
        if res is not None:
            return res
        visited.remove(ns)
    return None


# --- Salida ---

def bank_str(m, c):
    return f"{m}M, {c}C" if (m or c) else "-"


def format_state(state):
    m_left, c_left, boat = state
    m_right, c_right = 3 - m_left, 3 - c_left
    side = "IZQ" if boat == 0 else "DER"
    return f"IZQ: {bank_str(m_left, c_left)} | DER: {bank_str(m_right, c_right)} | Bote: {side}"


def describe_move(prev, move):
    if move is None:
        return "Estado inicial"
    m, c = move
    boat = prev[2]
    direction = "izquierda -> derecha" if boat == 0 else "izquierda <- derecha"
    parts = []
    if m:
        parts.append(f"{m} misionero(s)")
    if c:
        parts.append(f"{c} canibal(es)")
    who = " y ".join(parts)
    return f"El bote lleva {who} ({direction})"


def main():
    sol = dfs_backtracking()
    if not sol:
        print("No se encontró solución.")
        return
    print("Solución (backtracking):")
    print("Paso 0: Estado inicial")
    print(format_state(sol[0][0]))
    for i in range(1, len(sol)):
        prev_state = sol[i-1][0]
        state, move = sol[i]
        print(f"\nPaso {i}: {describe_move(prev_state, move)}")
        print(format_state(state))
    print(f"\nTotal de pasos: {len(sol)-1}")


if __name__ == "__main__":
    main()

File: Deber 2/ejercicio_4/mc_solver.pl
% mc_solver.pl - Misioneros y Caníbales con Backtracking (SWI-Prolog)
% Estado: state(M_left, C_left, Boat) con Boat en {l,r}

:- initialization(run, main).

% Estado inicial y objetivo
start(state(3,3,l)).
goal(state(0,0,r)).

% Lados opuestos
opp(l, r).
opp(r, l).

% Cargas posibles del bote (Misioneros, Caníbales)
boat_load(2,0).
boat_load(1,0).
boat_load(0,2).
boat_load(0,1).
boat_load(1,1).

% Un estado es seguro si ambas orillas respetan la regla
safe_counts(Ml, Cl) :-
    Ml >= 0, Ml =< 3, Cl >= 0, Cl =< 3,
    ( Ml =:= 0 ; Cl =< Ml ),             % izquierda
    Mr is 3 - Ml,
    Cr is 3 - Cl,
    ( Mr =:= 0 ; Cr =< Mr ).             % derecha

safe(state(Ml, Cl, _)) :- safe_counts(Ml, Cl).

% Movimiento: el bote siempre se mueve y lleva una carga válida
move(state(Ml, Cl, l), state(Ml2, Cl2, r), move(M,C)) :-
    boat_load(M, C),
    Ml >= M, Cl >= C,
    Ml2 is Ml - M, Cl2 is Cl - C,
    safe_counts(Ml2, Cl2).

move(state(Ml, Cl, r), state(Ml2, Cl2, l), move(M,C)) :-
    boat_load(M, C),
    Ml2 is Ml + M, Cl2 is Cl + C,
    Ml2 =< 3, Cl2 =< 3,
    safe_counts(Ml2, Cl2).

move(State, Next) :- move(State, Next, _).

% --- Backtracking (DFS) ---
% Encuentra una ruta cualquiera desde start a goal (no garantiza mínimos pasos)
solve(Path) :-
    start(S), goal(G),
    dfs(S, G, [S], Path).

dfs(State, Goal, _, [State]) :-
    State = Goal, !.

dfs(State, Goal, Visited, [State|Path]) :-
    move(State, Next),
    \+ member(Next, Visited),
    dfs(Next, Goal, [Next|Visited], Path).

% --- Salida ---

bank_str(0,0, '-') :- !.
bank_str(M, C, Str) :- format(string(Str), '~dM, ~dC', [M, C]).

print_state(state(Ml, Cl, _B)) :-
    Mr is 3 - Ml, Cr is 3 - Cl,
    bank_str(Ml, Cl, LStr),
    bank_str(Mr, Cr, RStr),
    format('IZQ: ~w | DER: ~w~n', [LStr, RStr]).

print_move(state(Ml1, Cl1, B1), state(Ml2, Cl2, _)) :-
    ( B1 == l -> Dir = 'izquierda -> derecha', M is Ml1 - Ml2, C is Cl1 - Cl2
    ;             Dir = 'izquierda <- derecha', M is Ml2 - Ml1, C is Cl2 - Cl1
    ),
    parts_str(M, C, Who),
    ( Who = '' -> format('El bote cruza vacio? (no permitido) (~w)~n', [Dir])
    ; format('El bote lleva ~w (~w)~n', [Who, Dir])
    ).

parts_str(M, C, Str) :-
    ( M =:= 0 -> MStr = '' ; format(string(MStr), '~d misionero(s)', [M]) ),
    ( C =:= 0 -> CStr = '' ; format(string(CStr), '~d canibal(es)', [C]) ),
    join_parts(MStr, CStr, Str).

join_parts('', '', '').
join_parts(A, '', A).
join_parts('', B, B).
join_parts(A, B, Str) :- format(string(Str), '~w y ~w', [A, B]).

print_solution([First|Rest]) :-
    writeln('Solucion (backtracking):'),
    writeln('Paso 0: Estado inicial'),
    print_state(First),
    print_steps(1, [First|Rest]).

print_steps(_, [_]).
print_steps(N, [Prev,Next|Rest]) :-
    format('~nPaso ~d: ', [N]),
    print_move(Prev, Next),
    print_state(Next),
    N1 is N+1,
    print_steps(N1, [Next|Rest]).

run :-
    ( solve(Path) ->
        print_solution(Path),
        length(Path, L), Steps is L-1,
        format('~nTotal de pasos: ~d~n', [Steps])
    ; writeln('No se encontró solución.')
    ).
